### ISSUE TITLE
API authentication fails with largish key sizes

### ISSUE THREAD

**Issue by mitom** [chef/supermarket#1656]
I am trying to set up a private supermarket for a private chef server. I have successfully deployed the supermarket and could authenticate to it without problems, as well as authorise the chef server. I can log in to the supermarket in the GUI without problems. However, I am unable to upload any cookbooks to it. When trying to upload one I just get:
```
knife cookbook site share test-cookbook Other
Generating metadata for test-cookbook from /var/folders/05/gywq5knd7kv2sy3tw9q931700000gp/T/chef-test-cookbook-build20170816-38701-wy3vox/test-cookbook/metadata.rb
Making tarball test-cookbook.tgz
ERROR: Authentication failed due to an invalid public/private key pair. If you have changed your keys recently try logging out and logging back in to Supermarket.
```

If I log in to the supermarket and look at my keys under my profile, the fingerprint matches the key's fingerprint (verified using the command detailed in the supermarket interface). 

I have synchronised the clock of the 2 servers, which did not change anything but was recommended in an earlier issue.

Tried looking at `chef-server-ctl tail oc_id` on the chef server, but it does not produce any logs when I try to share a cookbook. Also tried looking at the logs of sidekiq, rails and nginx on the supermarket. None of them produce any log entries. The only log that shows up is in the nginx access log which tells me I'm hitting the right server:
```
<hidden> - - [2017-08-16T12:40:31+00:00]  "POST /api/v1/cookbooks HTTP/1.1" 401 3722 "0.034" 226 "-" "Ruby"
```

Also tried logging out of the supermarket and the chef server as well and logging back in. Tried to cleanse the supermarket and reconfigured it to no avail.

I am completely out of ideas about what else to check, any help would be welcome. Let me know if any information is required.

**Comment by robbkidd** [comment]
Are you able to run other knife commands against the chef server—e.g. `knife node list`—to confirm that the knife configuration is using the expected credentials and key?

**Comment by mitom** [comment]
Yes I am. I have also tried changing my key on the chef server as well - it showed up correctly as a new fingerprint in the supermarket UI, but no luck. Have tried deleting the user completely and re-creating it, but no luck with that either.

Also destroyed the entire supermarket environment (database, instance and all) and started from scratch - also no luck. Interestingly, an other user from the same chef organisation which has identical setup seems to work without problems. I am not sure why, but my current guess is to dig through the chef server database for any oauth related things about my user and see if messing with those helps.

**Comment by robbkidd** [comment]
🤔  I'm going to go over some basics here. I apologize if these are things you already know, but maybe covering some of these things will trigger something that inspires another troubleshooting path.

Once you log in to the Supermarket UI, your username and public key is cached in Supermarket's database until you log out through the UI and log in again to refresh through oauth. API requests are authenticated with a signature created by the knife client with your private key and added to the request headers. The private key used by the client must match the _primary_ key associated with your user in the Chef server; Supermarket only caches and authenticates with the primary key, not all user keys.

I've seen a situation similar to yours when a user has multiple key pairs associated with their Chef user. Knife was configured to use one of the non-primary keys, so knife calls to the Chef Server would succeed and calls to Supermarket would fail as not auth'd.

**Comment by mitom** [comment]
No problem, it never hurts to cover all bases. 

Sadly, I have gone through this myself. Initially, I had multiple keys and the result was visible via the marketplace UI - the fingerprint for the key didn't match the one I was using. I simply removed all keys but the default, and replaced that with the desired one. Logged out and back in to supermarket and could verify that the fingerprint exactly matched my used key. I have also seen the cached key being updated properly when I tried to switch my key to a different one.
```
knife user key list my-username
default
```
```
knife user key show my-username default
expiration_date: infinity
name:            default
public_key:      -----BEGIN PUBLIC KEY-----
...(truncated)
roZxuN8soedrAyWKPXaTsUsCAwEAAQ==
-----END PUBLIC KEY-----
user:            my-username
```

What I'm suspicious of at the moment is, after deleting my user from the chef server and re-creating it, it retained the oauth authentication for the supermarket, which leads me to believe that it's not cleaned up properly. It appears that other users from the chef server can now simply join the supermarket and access it properly, including uploading cookbooks - except for me. My guess is because my user was associated to it in a very early state when it was not properly set up yet, there may be some remains of an invalid setting associated to my user somwhere..? It is a very wild guess, but I am not sure where else to look.

If you have any other ideas that I could look at, please let me know - even if they sound idiotic.

**Comment by mitom** [comment]
@robbkidd I may have found the issue, however I am uncertain about it. It appears that using a key size of 4kb is not working as the verification fails somewhere. My key is 4k of size (the original one) and when trying to generate a new pair, I used a 4k size as well. My collegue's key size is 2k and everything worked for him without issue. I have then generated myself a 2kb key and updated it in the same way as before, logged out/back in to supermarket - and it worked. I could use my new 2k key to upload to the supermarket. 

To double check that it was not just a fluke that some setting cleared it up or something else I did, I then generated yet another 4k key - which then produced the authentication error again.

Is there any way to double check for you if this is the case? Is there any such limitation? I doubt I'd be the first ever person to want to use a 4k size key on the supermarket.

**Comment by robbkidd** [comment]
> I then generated yet another 4k key - which then produced the authentication error again.

😱  Great sleuthing!

> Is there any way to double check for you if this is the case? Is there any such limitation?

I will investigate. If there is such a limitation, it's unintentional and a bug.

**Comment by robbkidd** [comment]
@mitom I slipped a 4096-bit key into a test Supermarket instance and was able to replicate the error.

**Comment by robbkidd** [comment]
Whittled it down to the library we're using for API request signature verification. The library has a change merged to master since its last release that fixes it. We'll release a new version of that gem and update Supermarket.

**Comment by robbkidd** [comment]
@mitom The fix is in the pipe. Great detective work. Thanks!

**Comment by mitom** [comment]
@robbkidd Thanks for the quick fix, could you give me an ETA when will this reach the stable release? It's a blocking issue for me at the moment.

**Comment by robbkidd** [comment]
@mitom The fix is in version [3.1.28](https://downloads.chef.io/supermarket/current/3.1.28), available at the moment in the `current` channel. I expect to let it bake in our staging environment over the weekend and release on Monday to stable. (Following the general rule of not releasing on a Friday.)
